kratos/mpi/utilities/mpi_search_utilities.h

```diff
@@ -0,0 +1,284 @@
+        if (AllPointsAreTheSame < 0) {
+            AllPointsAreTheSame = static_cast<int>(CheckAllPointsAreTheSame(itPointBegin, itPointEnd, NumberOfPoints, TotalNumberOfPoints, rDataCommunicator));
+        }
+
+        KRATOS_DEBUG_ERROR_IF(NumberOfPoints < 0) << "The number of points is negative" << std::endl;
+        KRATOS_DEBUG_ERROR_IF(TotalNumberOfPoints < 0) << "The total number of points is negative" << std::endl;
+
+        // If all points are the same
+        if (AllPointsAreTheSame) {
+            rAllPointsCoordinates.resize(NumberOfPoints * 3);
+            std::size_t counter = 0;
+            array_1d<double, 3> coordinates;
+            unsigned int i_coord;
+            for (auto it_point = itPointBegin ; it_point != itPointEnd ; it_point++) {
+                noalias(coordinates) = it_point->Coordinates();
+                for (i_coord = 0; i_coord < 3; ++i_coord) {
+                    rAllPointsCoordinates[3 * counter + i_coord] = coordinates[i_coord];
+                }
+                ++counter;
+            }
+        } else { // If not
+            // MPI information
+            const int world_size = rDataCommunicator.Size();
+
+            // Getting global number of points
+            std::vector<int> points_per_partition(world_size);
+            std::vector<int> send_points_per_partition(1, NumberOfPoints);
+            rDataCommunicator.AllGather(send_points_per_partition, points_per_partition);
+
+            // Getting global coordinates
+            rAllPointsCoordinates.resize(TotalNumberOfPoints * 3);
+            std::vector<double> send_points_coordinates(NumberOfPoints * 3);
+            std::size_t counter = 0;
+            array_1d<double, 3> coordinates;
+            unsigned int i_coord;
+            for (auto it_point = itPointBegin ; it_point != itPointEnd ; it_point++) {
+                noalias(coordinates) = it_point->Coordinates();
+                for (i_coord = 0; i_coord < 3; ++i_coord) {
+                    send_points_coordinates[3 * counter + i_coord] = coordinates[i_coord];
+                }
+                ++counter;
+            }
+
+            // Generate vectors with sizes for AllGatherv
+            std::vector<int> recv_sizes(world_size, 0);
+            for (int i_rank = 0; i_rank < world_size; ++i_rank) {
+                recv_sizes[i_rank] = 3 * points_per_partition[i_rank];
+            }
+            std::vector<int> recv_offsets(world_size, 0);
+            for (int i_rank = 1; i_rank < world_size; ++i_rank) {
+                recv_offsets[i_rank] = recv_offsets[i_rank - 1] + recv_sizes[i_rank - 1];
+            }
+
+            // Invoque AllGatherv
+            rDataCommunicator.AllGatherv(send_points_coordinates, rAllPointsCoordinates, recv_sizes, recv_offsets);
+        }
+
+        return NumberOfPoints;
+    }
+
+    /**
+     * @brief MPISynchronousPointSynchronization prepares synchronously the coordinates of the points for MPI search.
```

Comment:
This does nothing with the coordinates, only with the sizes. Note that, even if `rAllPointsCoordinates` is propagated during all the stack call, you only use `all_points_are_the_same`, `number_of_points` and `total_number_of_points`, which do not need from the coordinate transfer to be calculated.

Seems odd that the coordinate stuff is inside this function. 

Also it repeats a lot of code. This function is calling `CheckAllPointsAreTheSame` in line 150, but then in line 153 you are calling `MPISynchronousPointSynchronization` which again has the same call to `CheckAllPointsAreTheSame` inside.

Reply:
I think I will need to call you later, because I have questions